qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/snippets/QuteParentSnippetContext.java

```diff
@@ -0,0 +1,83 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.services.snippets;
+
+import java.io.IOException;
+import java.util.Map;
+
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+import com.redhat.qute.parser.template.Node;
+import com.redhat.qute.parser.template.NodeKind;
+import com.redhat.qute.parser.template.Section;
+import com.redhat.qute.services.completions.CompletionRequest;
+
+/**
+ * Snippet context for Qute sections that have parent sections.
+ *
+ */
+
+public class QuteParentSnippetContext implements IQuteSnippetContext {
+
+	public static final TypeAdapter<QuteParentSnippetContext> IN_PARENT = new SnippetContextForQuteParentAdapter();
+
+	private String parent;
+
+	public QuteParentSnippetContext(String parent) {
+		this.parent = parent;
+	}
+
+	@Override
+	public boolean isMatch(CompletionRequest request, Map<String, String> model) {
+		Node node = request.getNode();
+		if (node.getKind() == NodeKind.Section && node.getParentSection().getTag() == parent) {

```

Comment:
There is several problem with `node.getParentSection().getTag() == parent:`

 * node.getParentSection() can be null
 * node.getParentSection().getTag() can be null
 * equals should be used to compare String instead of ==
 * you don't take care of parent of parent, etc

I suggest that you move that in a private static method `isMatchParent(Node node, String parentTag)` :

```java
if (node.getKind() == NodeKind.Section) {
    Section section = (Section) node;
    if (!section.isInStartTagName(request.getOffset())) {
        return false;
    }
    return isMatchParent(node, this.parent);
}
```

isMatchParent could looks like this (not tested):

```
public static boolean isMatchParent(Node node, String parentTag) {
    Section parent = node.getParentSection();
    while (parent != null) {
        if (parentTag.equals(parent.getTag()) {
             return true;
        }
        parent = parent.getParentSection();
    }
    return false;
}
```

Reply:
I believe we only need to check the immediate parent, and not the nested one:

e.g. say we have
```
{#if condition1}
    {#if condition2}
        |
    {/if}
{/if}
```
the completion should only be for the if with condition2, I believe the if with condition1 doesn't matter here.

Reply:
> the completion should only be for the if with condition2, I believe the if with condition1 doesn't matter here.

Indeed you are right. Please update my suggestion with isMatchParent